sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs

```diff
@@ -98,41 +105,115 @@ public class EventProcessor<T> where T : BasePartitionProcessor, new()
         private Dictionary<string, PartitionOwnership> InstanceOwnership { get; set; }
 
         /// <summary>
-        ///   The running task responsible for performing partition load balancing between multiple <see cref="EventProcessor{T}" />
+        ///   The running task responsible for performing partition load balancing between multiple <see cref="EventProcessor" />
         ///   instances, as well as managing partition pumps and ownership.
         /// </summary>
         ///
         private Task RunningTask { get; set; }
 
         /// <summary>
-        ///   Initializes a new instance of the <see cref="EventProcessor{T}"/> class.
+        ///   The function to be called just before event processing starts for a given partition.
         /// </summary>
         ///
-        /// <param name="consumerGroup">The name of the consumer group this event processor is associated with.  Events are read in the context of this group.</param>
-        /// <param name="eventHubClient">The client used to interact with the Azure Event Hubs service.</param>
-        /// <param name="partitionManager">Interacts with the storage system with responsibility for creation of checkpoints and for ownership claim.</param>
-        /// <param name="options">The set of options to use for this event processor.</param>
+        public Func<PartitionContext, Task> InitializeProcessingForPartitionAsync
+        {
+            internal get => _initializeProcessingForPartitionAsync;
+
+            set
+            {
+                lock (_setHandlerMutex)
+                {
+                    if (RunningTask == null)
+                    {
+                        _initializeProcessingForPartitionAsync = value;
+                    }
+                    else
+                    {
+                        throw new InvalidOperationException(Resources.RunningEventProcessorCannotPerformOperation);
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        ///   The function to be called once event processing stops for a given partition.
+        /// </summary>
         ///
-        /// <remarks>
-        ///   Ownership of the <paramref name="eventHubClient" /> is assumed to be responsibility of the caller; this
-        ///   processor will delegate operations to it, but will not perform any clean-up tasks, such as closing or
-        ///   disposing of the instance.
-        /// </remarks>
+        public Func<PartitionContext, PartitionProcessorCloseReason, Task> ProcessingForPartitionStoppedAsync
+        {
+            internal get => _processingForPartitionStoppedAsync;
+
+            set
+            {
+                lock (_setHandlerMutex)
+                {
+                    if (RunningTask == null)
```

Comment:
If the `EventProcessor` encounters an unhandled error and stops running, we end up with `RunningTask != null` and the user explicitly needs to call `StopAsync` before being able to update the handlers again. That's intentional. Do you think it should behave differently?

Reply:
I do.   I think that the unhandled error trips up to the error handler and if that doesn't return `true` we stop processing and reset state.    Assuming, of course, that it's a processor level failure.

Reply:
That's something we can do when we update the handlers to match the new API (which should happen in the next PR). Right now it only returns a void `Task`. This PR is more focused in eliminating the need of a `PartitionProcessor`.